### DISCUSSION TITLE
How to handle FormData?

### DISCUSSION THREAD

**Discussion by srcnalt** [vercel/next.js#28943] [Help]
I am looking to find what is the best way to handle FormData with Next.js and wondering if `NextApiRequest` will support `fields and files` in the future versions.

Right now if I make a post request to my api end point, req.body looks like this for a string and a file:

```js
fetch(url, {
  method: "POST",
  body: formData,
  headers: { "Content-Type": "multipart/form-data" },
})
```

```
------WebKitFormBoundaryNBPBax5VnpaAWqRd
Content-Disposition: form-data; name="name"

John Doe
------WebKitFormBoundaryNBPBax5VnpaAWqRd
Content-Disposition: form-data; name="file"; filename="blob"
Content-Type: image/jpeg

����►JFIF☺☺☺☺��☻(ICC_PROFILE☺☺☻↑☻►mntrRGB XYZ acsp☺��☺�...
```

I want to access the string fields and the file of File, Blob or Buffer result from req, however thats currently not possible out of the box with Next.js afaik.

There are threads suggesting `formidable` and `multyparty` packages however could not get any of these work.

**Comment by wwwdepot** [comment]
You could use `formidable`, make sure turning off bodyParser in API route.

```js
// https://nextjs.org/docs/api-routes/api-middlewares#custom-config
export const config = {
  api: {
    bodyParser: false,
  },
}
```

**Reply by srcnalt** [reply]
I tried this but was not successful. I was asking this because I was trying to upoad files to S3, then realized there is a way to do it with signed URLs. So I decided to go that was and handle it client side.

**Comment by ogbotemi-2000** [comment]
In fact, I had a similar issue with popular packages for parsing multipart/form-data: _formidable, multer, multiparty_.
Given my deep knowledge of JavaScript regular expressions, I came up with a compact solution.

```js
function parseMultipart(req, data, bound, split) {
  /** not re-inventing the wheel or anything but popular packages like
   * formidable, multer and multiparty didn't work for my use case of sending
   * large text encoded as multipart/form-data hence the custom solution below
   */
  let body = {};
  req.headers['content-type'].replace(/=-+[^]+$/, match=>bound = match.replace(/=-+/, ''));
  split = new RegExp(`\\s*-*${bound}[^;]+(;|-+)\\s*`), data = data.split(split).filter(e=>e.length>1),
  data.forEach(e=>{
    let rgx = /^name="[^"]+"/ , value = e.split(rgx).filter(e=>e);
    e.replace(rgx, match=>body[match.replace(/name=|"/g, '')] = value.length ? value.shift().trim() : '')
  });
  return body;
}
```

>Binary data is stored as received and pointed to by the value of the `name` attribute of the input type (usually file), in the object. You may further use `Buffer` on it or write it to a file as-is
